### PR TITLE
Implement time-predicate support

### PR DESCRIPTION
**Pull request by dburkart**
This PR:

  * Implements time-predicates
  * Improves sample() to take a time-quantity (instead of just a timespan)
  * Fixes a few retrieval accuracy bugs, [see this commit](https://github.com/dburkart/fossil/pull/39/commits/03bcf74118d5f699a0ecdee5677108019697903f) for more context
  * Changes the grammar slightly for usability reasons

### CODE REVIEW

File: pkg/query/ast.go [hunk 1]
```diff
@@ -18,10 +18,46 @@ type ASTNode interface {
 	Walk(*database.Database) []database.Filter
 }
 
-type BaseNode struct {
-	Value    string
-	children []ASTNode
-}
+type (
```
Comment (gideonw) [inline pkg/query/ast.go:21]: You did the thing!

Comment (dburkart) [reply pkg/query/ast.go:21]: Yeah! It feels more organized

**Comment by dburkart** [comment]
Cool little usage example:

```
> QUERY all
2022-12-02 10:05:10.9999 -0800 PST	/	foo
2022-12-02 10:06:37.57749 -0800 PST	/	bar
2022-12-02 10:09:10.67357 -0800 PST	/	baz
2022-12-02 10:09:37.12085 -0800 PST	/	blah


> QUERY sample(@minute) since ~2022-12-02T10:06:37-08:00
2022-12-02 10:06:37.57749 -0800 PST	/	bar
2022-12-02 10:09:10.67357 -0800 PST	/	baz


> QUERY sample(@second) since ~2022-12-02T10:06:37-08:00
2022-12-02 10:06:37.57749 -0800 PST	/	bar
2022-12-02 10:09:10.67357 -0800 PST	/	baz
2022-12-02 10:09:37.12085 -0800 PST	/	blah
```

**Comment by dburkart** [comment]
Before example:

```
> QUERY all                                  
2022-12-02 10:05:10.9999 -0800 PST	/	foo
2022-12-02 10:06:37.57749 -0800 PST	/	bar
2022-12-02 10:09:10.67357 -0800 PST	/	baz
2022-12-02 10:09:37.12085 -0800 PST	/	blah


> QUERY all before ~2022-12-02T10:06:37-08:00
2022-12-02 10:05:10.9999 -0800 PST	/	foo
```

**Comment by dburkart** [comment]
Cool:

```
> QUERY all since ~now - @day * 2
No Results

> QUERY all since ~now - @day * 3
2022-11-30 08:19:40.130813014 -0800 PST	/	it is now day 2, and this is my 7th event
2022-11-30 12:27:43.002961753 -0800 PST	/	the database now has a STAT command :^)
2022-11-30 22:18:12.635156609 -0800 PST	/	Sending data to iotas (1)
2022-11-30 22:18:12.684801384 -0800 PST	/	Sending data to iotas (2)
2022-11-30 22:18:12.711650419 -0800 PST	/	Sending data to iotas (3)
2022-11-30 22:18:12.74364688 -0800 PST	/	Sending data to iotas (4)
2022-11-30 22:18:12.772995624 -0800 PST	/	Sending data to iotas (5)
2022-11-30 22:18:12.80093928 -0800 PST	/	Sending data to iotas (6)
2022-11-30 22:18:12.830923059 -0800 PST	/	Sending data to iotas (7)
2022-11-30 22:18:12.862755216 -0800 PST	/	Sending data to iotas (8)
2022-11-30 22:18:12.89082536 -0800 PST	/	Sending data to iotas (9)


> QUERY all before ~now - @day * 3
2022-11-29 21:52:30.899373572 -0800 PST	/	Welcome to iotas.local!
2022-11-29 22:20:31.711798581 -0800 PST	/	Hooray! Fossil is running as a daemon on iotas.local!
2022-11-29 22:30:06.67726485 -0800 PST	/	Let's see how long this thing will stay up
2022-11-29 23:19:19.105794922 -0800 PST	/	testing from a pipe (1)
2022-11-29 23:19:19.109712759 -0800 PST	/	testing from a pipe (2)
```

And sampling gets time-quantities too!

```
> QUERY sample(@hour)
2022-11-29 21:52:30.899373572 -0800 PST	/	Welcome to iotas.local!
2022-11-29 23:19:19.105794922 -0800 PST	/	testing from a pipe (1)
2022-11-30 08:19:40.130813014 -0800 PST	/	it is now day 2, and this is my 7th event
2022-11-30 12:27:43.002961753 -0800 PST	/	the database now has a STAT command :^)
2022-11-30 22:18:12.635156609 -0800 PST	/	Sending data to iotas (1)


> QUERY sample(@minute * 10)
2022-11-29 21:52:30.899373572 -0800 PST	/	Welcome to iotas.local!
2022-11-29 22:20:31.711798581 -0800 PST	/	Hooray! Fossil is running as a daemon on iotas.local!
2022-11-29 23:19:19.105794922 -0800 PST	/	testing from a pipe (1)
2022-11-30 08:19:40.130813014 -0800 PST	/	it is now day 2, and this is my 7th event
2022-11-30 12:27:43.002961753 -0800 PST	/	the database now has a STAT command :^)
2022-11-30 22:18:12.635156609 -0800 PST	/	Sending data to iotas (1)
```

**Comment by dburkart** [comment]
And timestamps are now specified with parenthesis:

```
> QUERY all since ~(2022-11-30T22:00:00-08:00)
2022-11-30 22:18:12.635156609 -0800 PST	/	Sending data to iotas (1)
2022-11-30 22:18:12.684801384 -0800 PST	/	Sending data to iotas (2)
2022-11-30 22:18:12.711650419 -0800 PST	/	Sending data to iotas (3)
2022-11-30 22:18:12.74364688 -0800 PST	/	Sending data to iotas (4)
2022-11-30 22:18:12.772995624 -0800 PST	/	Sending data to iotas (5)
2022-11-30 22:18:12.80093928 -0800 PST	/	Sending data to iotas (6)
2022-11-30 22:18:12.830923059 -0800 PST	/	Sending data to iotas (7)
2022-11-30 22:18:12.862755216 -0800 PST	/	Sending data to iotas (8)
2022-11-30 22:18:12.89082536 -0800 PST	/	Sending data to iotas (9)
```

File: pkg/query/ast.go [hunk 2]
```diff
@@ -147,28 +191,121 @@ func (q TopicSelectorNode) GenerateFilter(db *database.Database) database.Filter
...
+	panic(fmt.Sprintf("Unknown operator '%s'", b.Value))
+}
+
+//-- TimespanNode
+
+func (t TimespanNode) DerivedValue() int64 {
 	switch t.Value {
 	case "@year":
-		return time.Hour * 24 * 365
+		return int64(time.Hour * 24 * 365)
```
Comment (gideonw) [inline pkg/query/ast.go:288]: What does this effectively use as a unit? int64 seconds? ms? ns?

**Review by gideonw — APPROVED** [review]
LGTM, one question for understanding.

File: pkg/query/ast.go (hunk 2 above)
Comment (dburkart) [reply pkg/query/ast.go:288]: `time.Duration` are nanoseconds, and is an `int64` under the hood... I wanted an interface so either had to convert everything to a duration, or convert these back to `in64`. I am not sure I made the right decision

**Pull request merged**